compiler/reports.nim

```diff
@@ -0,0 +1,108 @@
+## This module provides type definitions for all structured report entries
+## that compiler can provide.
+##
+## Note that this module specifically does not import anything else from
```

Comment:
Main use case for the import-ast solution is reporting of the overload resolution errors - "but expression has type", known proc signatures.

Reply:
`msgs.writeContex()` (and some other procedures) report information in form of `"XXX of <name of the procedure/template/macro>`, maybe with `defined in <location>`. Report might contain type like 

```nim
SemReportEntry = object
  name*: string
  kind*: enum = sreProc, sreMacro, sreField, sreXXX ...
  definedIn*: ReportLineInfo
  sigDescription*: <object that encodes whole description of the used entry signature>
                   <for example for procs it would be an arg/return pragmas and so on>
```

If ast is used we would need to reconstruct the whole "sig description" report using procs that traverse the AST and get necessary information for pretty printing, and this kind of thing would be very annoying to serialize properly.

Reply:
I think it is better to decouple these sections of the compiler at the cost of some repetition, since we would still have to do *exactly the same* thing for pretty-printing or serialization, so why not prepare the data fully beforehand.

Reply:
`m*: MsgConfig` from `ConfigRef` is populated via procs like 

```nim
proc pushInfoContext*(conf: ConfigRef; info: TLineInfo; detail: string = "") =
  conf.m.msgContext.add((info, detail))
```

which in turn are populated via `sym.name` (sorry, `template detailedInfo*(sym: PSym): string = sym.name.s`, so we would need to push syms into message context for the config anyway)

Reply:
```nim
  SemReportEntryKind* = enum
    srekProc
    srekTemplate
    srekMacro
    srekMethod
    srekConverter

    srekEnum
    srekObject
    srekAlias
    srekTypeclass

  SemReportEntry* = object
    ## Entry mentioned in the sem report - type or procedure definition,
    ## macros, template or any similar construct. Used in reports like
    ## overload failures, "template/generic instantiation of" and such, to
    ## store information about entry referred to.
    declaredIn*: ReportLineInfo ## Location of the entry declaration
    kind*: SemReportEntryKind
```

Pushed in the b395e02

Reply:
Is it safe to push syms into configRef `m*: MsgConfig` (`msgContext*: seq[tuple[info: TLineInfo, detail: string]]`) and then later unpack things from there as needed?

Reply:
Modularity and data oriented go hand in hand I find. Basically the nkError stuff bright back a whole bunch of old lessons for me confined with a bunch of other old lessons from seemingly unrelated experience. I'll speak generally first, then try and go more specific.

----

### general

Generally speaking I think module decomposition should work like so:
1. Write a module that covers a while topic or responsibility and call it good

Once the dependencies start arising out they're there from the get go, because existing code the fuller decomposition of modules is more like this:
- data module: fundamental data for the topic, optimized for write or reporting of that's much more expensive
- command or write module: contains the constructors, necessary or minimal read functionality for queries required for write decision making, traversal, etc; all the sort of create commands
- effects or reporting or read module: all the heavy reads and consequent actions, uses the above module to traverse and decide where necessary; might produce secondary or additional data structures related to its bookkeeping

The various dependencies force a separation like this anyhow because otherwise you end up with recursive import issues.

So now back to this, the current AST isn't data oriented, but also not entirely the worst, and this has a mild cascade effect that it as some chonky boys like ConfigRef and friends to come along for the ride. The abstractions to work around this aren't great either.

----

### specific

I'm always more partial to fundamental fixes, which I think would begin with wrapping all node construction with a config aware constructor. That would let us use the config as a context which will enable making the whole AST data oriented. Locating all the construction sites might not be too bad, if we can leverage error or deprecation pragma or some other trickery.

If the above feels like a mountain then I would head the data duplication route, but make this data oriented so we can more easily converge later, somehow.

Reply:
I can't exactly go into fundamental fixes there, because it would create this crooked intermediate state where I spend just as much time making work continiously instead of burning this whole contraption to the ground. Example - "template/generic instantiation of" is built on top of the construction that essentially appears to be a stack of pairs that are push-pop edited via `msgs.pushInfoContext` and `msgs.popInfoContext`. This stack contains strings, or "detail" information, which is generated immediately in various places. So I need to get rid of this string as well, otherwise I'm unable to even construct `SemContext` object meaningfully. I could've just made it into a string and passed it along, but this feels like an absolute hack that I would have to remove anyway.

Reply:
> but make this data oriented so we can more easily converge later, somehow. 

Can you elaborate on this one specifically - does my approach seem like a move in a right direction for that, or I need to change something? Because right now I just want to go in the "Create an object that thoroughly describes the report" route. Later this object is simply converter to a string (for pretty-printing) or json (for structured reporting)

Reply:
> 

Yes, I believe this is safe, the only time things get a bit weird is that we could be in metaprogramming land, like a compilation inside a compilation, and so the inner stuff might be throw away. If we hang onto a ref of the symbols from the inner speculative compilation then we will have potential space leak (that the right term?).

Reply:
> Can you elaborate on this one specifically - does my approach seem like a move in a right direction for that, or I need to change something? Because right now I just want to go in the "Create an object that thoroughly describes the report" route. Later this object is simply converter to a string (for pretty-printing) or json (for structured reporting)

From a strict performance perspective it isn't. I don't care about that and it totally misses the primary benefit of data oriented design.

I'm looking at the data and seeing about normalizing it, mostly. You've got no refs which is a good sign. The only thing that stands out is the inheritance, which isn't always a non-starter, and I'd consider the look aside arrays/tables approach, but I'm not a purist. I see it potentially needing to change but I'll admit this might be easier to read.

For me I'd say yes for now, but watch out if you feel the urge to add refs in the middle of the data.